### DISCUSSION TITLE
How to extract values from TSpline?

### DISCUSSION THREAD

**Discussion by yixianz** [scikit-hep/uproot5#1275] [Q&A]
Hi! I'm new to uproot so maybe I'm missing something obvious. 
I was trying to figure out how I can extract values from TSpline3. In ROOT there are functions like Eval(), FindX(), etc; what's the equivalent way of doing this in uproot?

**Comment by jpivarski** [comment]
The short answer is that any TSpline3 objects read in with Uproot don't have methods, only data members. In Python, we have the ROOT file and therefore all of the data, but not the C++ code. So there's no way to do this now.

The longer answer is that methods like this can be added to the `Model_TSpline3_vN` class by adding a "behavior" in https://github.com/scikit-hep/uproot5/tree/main/src/uproot/behaviors . It involves implementing these methods in Python, operating on the data that's already there.

The first step, whether you're thinking about contributing these functions or just getting something to work for yourself, is to look at the `all_members` and access individual members by `member("fMemberName")` to see the data that are there, and then think about what you need to compute.

**Reply by yixianz** [reply]
Thank you for the suggestion! I looked at the `all_members` and could see some data there (as shown below), but I might still need some help moving forward. 
```
'@fUniqueID': 0,
 '@fBits': 16777216,
 'fName': 'spline10',
 'fTitle': 'spline10',
 'fLineColor': 1,
 'fLineStyle': 1,
 'fLineWidth': 1,
 'fFillColor': 0,
 'fFillStyle': 1,
 'fMarkerColor': 1,
 'fMarkerStyle': 1,
 'fMarkerSize': 1.0,
 'fDelta': -1.0,
 'fXmin': -57.7689,
 'fXmax': 1061.0729999999999,
 'fNp': 13,
 'fKstep': False,
 'fHistogram': None,
 'fGraph': None,
 'fNpx': 100,
 'fPoly': <TSplinePoly3 (version 1) at 0x7fefd20f4fd0>,
 'fValBeg': 0.0,
 'fValEnd': 0.0,
 'fBegCond': 2,
 'fEndCond': 2
```
For my project at least I would like to get the knot values and there should be 13 knots in this case. In principle, `fPoly` should have them, which was defined as "[fNp] Array of polynomial terms" in ROOT. But if I do `member('fPoly').all_members`, I just get:
```
'@fUniqueID': 0,
 '@fBits': 16777216,
 'fX': 1061.0729999999999,
 'fY': 83.97380000000001,
 'fB': 0.0759236684030201,
 'fC': 321.9889999999999,
 'fD': 1.744866466793852
```
which I believe is just showing results for one knot. Am I missing something here?

**Reply by jpivarski** [reply]
I wonder if the `fPoly` is an array, TSplinePoly3*`, but Uproot is reading it as though it was only one instance (interpreting the pointer as though it were an item, rather than an array). If that were the case, then the next value would be wrongly interpreted as `fValBeg`, `fValEnd`, `fBegCond`, `fEndCond`. But then, it would also fail the test the checks the number of expected bytes against the number of observed bytes, and you would have gotten an error message instead of an object like this, so that's probably not it. 

What does

```python
uproot.open("file.root").file.show_streamers()
```

show?

**Reply by jpivarski** [reply]
https://uproot.readthedocs.io/en/latest/uproot.reading.ReadOnlyFile.html#show-streamers

**Reply by yixianz** [reply]
The output is really long; I'm just copying relevant pieces here:
```
TSpline3 (v2): TSpline (v2)
    fPoly: TSplinePoly3* (TStreamerLoop)
    fValBeg: double (TStreamerBasicType)
    fValEnd: double (TStreamerBasicType)
    fBegCond: int (TStreamerBasicType)
    fEndCond: int (TStreamerBasicType)

TSpline (v2): TNamed (v1), TAttLine (v2), TAttFill (v2), TAttMarker (v2)
    fDelta: double (TStreamerBasicType)
    fXmin: double (TStreamerBasicType)
    fXmax: double (TStreamerBasicType)
    fNp: int (TStreamerBasicType)
    fKstep: bool (TStreamerBasicType)
    fHistogram: TH1F* (TStreamerObjectPointer)
    fGraph: TGraph* (TStreamerObjectPointer)
    fNpx: int (TStreamerBasicType)

... 

TSplinePoly3 (v1): TSplinePoly (v2)
    fB: double (TStreamerBasicType)
    fC: double (TStreamerBasicType)
    fD: double (TStreamerBasicType)

TSplinePoly (v2): TObject (v1)
    fX: double (TStreamerBasicType)
    fY: double (TStreamerBasicType)

```

**Reply by jpivarski** [reply]
It's probably the `TStreamerLoop`, which is a more obscure TStreamer type and may not be implemented correctly. (But then, I wonder why the object didn't fail deserialization with a message about having the wrong number of bytes: misinterpreting 13 `TSplinePoly3` as 1 `TSplinePoly3` would set the number of bytes read incorrectly.)

I can take a look at this if you supply a small file that can be accessed publicly (i.e. it shouldn't contain any private data). I can use it to fix the `TStreamerLoop` implementation, if that's what the problem is, and then include it in our automated testing for the future.

**Reply by yixianz** [reply]
That would be great! Thank you for your help. 
I made a simple TSpline3 with 3 knots (x = [1,2,3], y = x^2), which should be accessible from this link:
https://drive.google.com/file/d/1ZcLWr1_udVZLNZne4GfWtDlqazUERsFT/view?usp=drive_link.

**Reply by jpivarski** [reply]
This should be fixed in #1277. Let me know if it works for you on all of your files, and if so, I'll merge it in and it will make its way into the next release.

**Reply by yixianz** [reply]
Yes, it works! Thank you!
